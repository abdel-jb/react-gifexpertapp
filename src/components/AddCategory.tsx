import React from 'react';

type TAddCategory = {
    setCategories: React.Dispatch<React.SetStateAction<string[]>> | any;
}

export const AddCategory = ({ setCategories }: TAddCategory): JSX.Element => {

    const [inputValue, setInputValue] = React.useState('');

    const handleInputChanhe = (e: React.FormEvent<HTMLInputElement>): void => {
        const { value } = e.target as HTMLInputElement;
        setInputValue(value);
    }

    const handleSubmit = (e: React.FormEvent): void => {
        e.preventDefault();
        if (inputValue.trim().length > 2) {
            setCategories(( category: string[] ) => (!category.find(e => e === inputValue) ? [inputValue, ...category] : category));
            setInputValue('');
        }
    };

    return (
        <form action="" onSubmit={ handleSubmit }>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChanhe }
            />
        </form>
    )
}
