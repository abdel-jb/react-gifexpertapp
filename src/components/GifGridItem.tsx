import React from 'react';

import { TGifs } from '../utils/types';


export const GifGridItem = ( { title, url }: TGifs ): JSX.Element => {
    return (
        <li className="card animate__animated animate__fadeIn">
            <img src={ url } alt={ title }/>
            <p>{ title }</p>
        </li>
    )
}
