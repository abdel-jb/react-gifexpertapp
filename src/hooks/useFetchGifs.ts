import React from 'react';
import { getGifs } from '../helpers/getGifs';

type TUseFetchGifsState = {
    data: any[],
    loading: boolean 
}

export const useFetchGifs = ( category: string ) => {
    const [state, setState] = React.useState<TUseFetchGifsState>({
        data: [],
        loading: true
    });
    
    React.useEffect(() =>{
        getGifs(category)
            .then((imgs) => {
                setState({
                    data: imgs,
                    loading: false
                });
        });
    }, [category]);

    return state;
}