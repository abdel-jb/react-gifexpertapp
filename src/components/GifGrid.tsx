import React from 'react';

import { GifGridItem } from './GifGridItem';
import { TGifs } from '../utils/types';
import { useFetchGifs } from '../hooks/useFetchGifs';


type TGifGrid = {
    category: string
}

export const GifGrid = ({ category }: TGifGrid): JSX.Element => {

    const { data: images, loading } = useFetchGifs(category);

    return (
        <>
            <h3>{ category }</h3>

            { loading && <p>Loading...</p> }

            <ul className="card-grid">
                { images.map(( img: TGifs ) => 
                    <GifGridItem
                        key={ img.id }
                        { ...img }
                    />
                )}
            </ul>
        </>
    )
}
