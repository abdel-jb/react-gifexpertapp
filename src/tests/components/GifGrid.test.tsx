import React from 'react';
import { shallow } from 'enzyme';

import { useFetchGifs } from '../../hooks/useFetchGifs';
import { GifGrid } from '../../components/GifGrid';

jest.mock('../../hooks/useFetchGifs');

describe('Testing GifGrid', () => {

    const category = 'One Punch';
    
    test('should display correctly', () => {
        
        // @ts-ignore
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={ category }/>);
        
        expect( wrapper ).toMatchSnapshot();
    
    });
    
    test('should show items when loading images useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/test.jpg',
            title: 'Test'
        }];

        // @ts-ignore
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: true
        });


        const wrapper = shallow(<GifGrid category={ category }/>);

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe( true );
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );
        
    });
    
});
