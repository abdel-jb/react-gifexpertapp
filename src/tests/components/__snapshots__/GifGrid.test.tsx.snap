// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGrid should display correctly 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <p>
    Loading...
  </p>
  <ul
    className="card-grid"
  />
</Fragment>
`;

exports[`Testing GifGrid should show items when loading images useFetchGifs 1`] = `
<Fragment>
  <h3>
    One Punch
  </h3>
  <p>
    Loading...
  </p>
  <ul
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Test"
      url="https://localhost/test.jpg"
    />
  </ul>
</Fragment>
`;
